src/lib/comms/sol-connman-impl-connman.c

```diff
@@ -0,0 +1,910 @@
+    SOL_INT_CHECK(r, < 0, r);
+
+    return 0;
+}
+
+int
+sol_connman_init(void)
+{
+    return 0;
+}
+
+void
+sol_connman_shutdown(void)
+{
+    struct sol_connman_service *service;
+    uint16_t id;
+
+    if (_ctx.connman) {
+        sol_bus_client_free(_ctx.connman);
+        _ctx.connman = NULL;
+    }
+
+    _ctx.state_slot =
+        sd_bus_slot_unref(_ctx.state_slot);
+    _ctx.manager_slot =
+        sd_bus_slot_unref(_ctx.manager_slot);
+    _ctx.service_slot =
+        sd_bus_slot_unref(_ctx.service_slot);
+
+    SOL_VECTOR_FOREACH_IDX (&_ctx.service_vector, service, id) {
+        _free_connman_service(service);
+    }
+
+    sol_vector_clear(&_ctx.service_vector);
+    sol_vector_clear(&_ctx.monitor_vector);
+
+    _ctx.connman_state = SOL_CONNMAN_STATE_UNKNOWN;
+}
+
+static int
+sol_connman_init_lazy(void)
+{
+    sd_bus *bus;
+
+    if (_ctx.connman)
+        return 0;
+
+    bus = sol_bus_get(NULL);
+    if (!bus) {
+        SOL_WRN("Unable to get sd bus\n");
+        return -EINVAL;
+    }
+
+    _ctx.connman = sol_bus_client_new(bus, "org.connman");
+    if (!_ctx.connman) {
+        SOL_WRN("Unable to new a bus client\n");
+        return -EINVAL;
+    }
+
+    sol_vector_init(&_ctx.service_vector, sizeof(struct sol_connman_service));
+    sol_vector_init(&_ctx.monitor_vector, sizeof(struct dbus_monitor_callback));
+
+    return 0;
+}
+
+static void
+sol_connman_shutdown_lazy(void)
+{
+    struct sol_connman_service *service;
+    uint16_t id;
+
+    if (_ctx.connman) {
+        sol_bus_client_free(_ctx.connman);
+        _ctx.connman = NULL;
+    }
+
+    _ctx.state_slot =
+        sd_bus_slot_unref(_ctx.state_slot);
+    _ctx.manager_slot =
+        sd_bus_slot_unref(_ctx.manager_slot);
+    _ctx.service_slot =
+        sd_bus_slot_unref(_ctx.service_slot);
+
+    SOL_VECTOR_FOREACH_IDX (&_ctx.service_vector, service, id) {
+        _free_connman_service(service);
+    }
+
+    sol_vector_clear(&_ctx.service_vector);
+    sol_vector_clear(&_ctx.monitor_vector);
+
+    _ctx.connman_state = SOL_CONNMAN_STATE_UNKNOWN;
+}
+
+static int
+_match_properties_changed(sd_bus_message *m, void *userdata,
+    sd_bus_error *ret_error)
+{
+    const char *interface;
+    int r;
+
+    SOL_NULL_CHECK(_ctx.properties_changed, -EINVAL);
+
+    interface = sd_bus_message_get_interface(m);
+    SOL_NULL_CHECK(interface, -EINVAL);
+
+    if (strncmp(interface, "net.connman.", 12) != 0)
+        return -EINVAL;
+
+    if (sd_bus_message_is_signal(m, "net.connman.Manager",
+        "ServicesChanged")) {
+        if (_ctx.service_slot)
+            return -EINVAL;
+
+        services_list_changed(m);
+    } else if (sd_bus_message_is_signal(m, "net.connman.Manager",
+        "PropertyChanged")) {
+        char *str = NULL;
+
+        if (_ctx.manager_slot)
+            return -EINVAL;
+
+        r = sd_bus_message_read_basic(m, SD_BUS_TYPE_STRING, &str);
+        SOL_INT_CHECK(r, < 0, r);
+        if (strcmp(str, "State") == 0) {
```

Comment:
as for `services_list_changed()`, couldn't you abstract this to a common function shared with the GetProperties?